PetaPoco.Tests.Integration/Databases/MySQL/MySqlQueryTests.cs

```diff
@@ -9,5 +15,169 @@ public MySqlQueryTests()
+                         ORDER BY 1 DESC
+                         LIMIT 1;";
+
+            List<Order> result;
+            using (var multi = DB.QueryMultiple(sql, "Peta0"))
+            {
+                result = multi.Read<Order, Person, Order>((o, p) =>
+                {
+                    o.Person = p;
+                    return o;
+                }).ToList();
+            }
+
+            result.Count.ShouldBe(1);
+
+            var order = result.First();
+
+            order.PoNumber.ShouldStartWith("PO");
+            order.Status.ShouldBeOneOf(Enum.GetValues(typeof(OrderStatus)).Cast<OrderStatus>().ToArray());
+            order.PersonId.ShouldNotBe(Guid.Empty);
+            order.CreatedOn.ShouldBeLessThanOrEqualTo(new DateTime(1990, 1, 1, 0, 0, 0, DateTimeKind.Utc));
+            order.CreatedBy.ShouldStartWith("Harry");
+
+            order.Person.ShouldNotBeNull();
+            order.Person.Id.ShouldNotBe(Guid.Empty);
+            order.Person.Name.ShouldStartWith("Peta");
+            order.Person.Age.ShouldBe(18);
+        }
+
+        [Fact]
+        public override void QueryMultiple_ForMultiResultsSetWithSinglePoco_ShouldReturnValidPocoCollection()
+        {
+            AddOrders(1);
+
+            var pd = PocoData.ForType(typeof(Person), DB.DefaultMapper);
+            var od = PocoData.ForType(typeof(Order), DB.DefaultMapper);
+            var pdName = pd.Columns.Values.First(c => c.PropertyInfo.Name == "Name").ColumnName;
+            var odId = od.Columns.Values.First(c => c.PropertyInfo.Name == "Id").ColumnName;
+
+            var sql = $@"SELECT * FROM {DB.Provider.EscapeTableName(od.TableInfo.TableName)} o
+                         WHERE o.{DB.Provider.EscapeSqlIdentifier(odId)} = @0;
+                         SELECT * FROM {DB.Provider.EscapeTableName(pd.TableInfo.TableName)} p
+                         WHERE p.{DB.Provider.EscapeSqlIdentifier(pdName)} = @1;";
+
+            Order order;
+            using (var multi = DB.QueryMultiple(sql, "1", "Peta0"))
+            {
+                order = multi.Read<Order>().First();
+                order.Person = multi.Read<Person>().First();
+            }
+
+            order.PoNumber.ShouldStartWith("PO");
+            order.Status.ShouldBeOneOf(Enum.GetValues(typeof(OrderStatus)).Cast<OrderStatus>().ToArray());
+            order.PersonId.ShouldNotBe(Guid.Empty);
+            order.CreatedOn.ShouldBeLessThanOrEqualTo(new DateTime(1990, 1, 1, 0, 0, 0, DateTimeKind.Utc));
+            order.CreatedBy.ShouldStartWith("Harry");
+
+            order.Person.ShouldNotBeNull();
+            order.Person.Id.ShouldNotBe(Guid.Empty);
+            order.Person.Name.ShouldStartWith("Peta");
+            order.Person.Age.ShouldBe(18);
+        }
+
+        // FIXME: firstOrderLine.Quantity.ToString() should be "1" but was "2"
+        [Fact]
+        public override void QueryMultiple_ForMultiResultsSetWithMultiPoco_ShouldReturnValidPocoCollection()
```

Comment:
Seems this should be obvious, but I'm missing what's wrong here. Same assertion failure manifesting on all MySql-based dbms and drivers (MySql, MySql Connector, MariaDB)

Reply:
I'm not sure *how* this could be happening, but this test passes on the MySQL instance in my local dev env (v8.0.25). In the Docker image (latest: v8.1.0), it fails. Manually inspecting a few records from `OrderLines`:

```sql
`SELECT * FROM `OrderLines` ol ORDER BY ol.`OrderId` ASC;`
```

| # Id | OrderId | Qty | Status | SellPrice |
|------|---------|-----|--------|-----------|
| '1'  | '1'     | '1' | '0'    | '9.9900'  |
| '2'  | '1'     | '2' | '0'    | '19.9800' |
| '3'  | '2'     | '1' | '0'    | '9.9900'  |
| '4'  | '2'     | '2' | '0'    | '19.9800' |

It occurred to me since we're only sorting based on `OrderId`, and since there are two records for each unique `OrderId`, it would be possible I guess that a valid ordering could also be:

| # Id | OrderId | Qty | Status | SellPrice |
|------|---------|-----|--------|-----------|
| '2'  | '1'     | '2' | '0'    | '19.9800' |
| '1'  | '1'     | '1' | '0'    | '9.9900'  |
| '4'  | '2'     | '2' | '0'    | '19.9800' |
| '3'  | '2'     | '1' | '0'    | '9.9900'  |

Which, technically, isn't wrong, but most certainly does qualify as undefined behavior.

I'm still mystified as to why it's only exhibited on MySQL DBMS on the Docker image, and not my local (or the other DBMS instances on Docker either, for that behavior).

Tests now pass, with a modified ORDER BY:

```sql
SELECT * FROM `OrderLines` ol ORDER BY ol.`Id` ASC, ol.`OrderId` ASC;
```

Still rather unsettling as I don't know what's causing the sorting to consistently differ between db instances.

https://github.com/CollaboratingPlatypus/PetaPoco/blob/c2dbb11c8fc5c10938f577acbcc57b22761b9fd7/PetaPoco.Tests.Integration/Databases/MySQLConnector/MySqlConnectorQueryTests.cs#L140